File: .github/scripts/build-pr-payload.js
#!/usr/bin/env node
'use strict';

/**
 * Builds the JSON payload for the typst/packages pull request.
 *
 * Reads the relevant CHANGELOG.md section (if present) and writes
 * /tmp/pr_payload.json for use with `gh api repos/typst/packages/pulls`.
 *
 * Required environment variables:
 *   PACKAGE_NAME  e.g. cgc-resume-starter
 *   VERSION       e.g. 2.0.0
 *   BRANCH        e.g. add/cgc-resume-starter-2.0.0
 */

'use strict';

const { readFileSync, writeFileSync, existsSync } = require('node:fs');

function extractChangelogEntry(packageName, version) {
  const changelogPath = `packages/${packageName}/CHANGELOG.md`;
  if (!existsSync(changelogPath)) return '';

  const content = readFileSync(changelogPath, 'utf8');
  const escapedVersion = version.replace(/\./g, '\\.');
  const headingRe = new RegExp(`^## .*${escapedVersion}.*$`, 'm');
  const headingMatch = headingRe.exec(content);
  if (!headingMatch) return '';

  const start = headingMatch.index + headingMatch[0].length;
  const nextHeading = /^## /m.exec(content.slice(start));
  const end = nextHeading ? start + nextHeading.index : content.length;
  return content.slice(start, end).trim();
}

function main() {
  const packageName = process.env.PACKAGE_NAME ?? '';
  const version = process.env.VERSION ?? '';
  const branch = process.env.BRANCH ?? '';

  if (!packageName || !version || !branch) {
    console.error('::error::PACKAGE_NAME, VERSION, and BRANCH must be set');
    process.exit(1);
  }

  const body = extractChangelogEntry(packageName, version);
  const payload = {
    title: `packages/preview/${packageName}/${version}`,
    head: `chaoticgoodcomputing:${branch}`,
    base: 'main',
    body,
  };

  writeFileSync('/tmp/pr_payload.json', JSON.stringify(payload));
  console.log('PR payload written to /tmp/pr_payload.json');
  console.log(`  title : ${payload.title}`);
  console.log(`  head  : ${payload.head}`);
  console.log(`  body  : ${body.length} chars`);
}

main();

File: .github/scripts/build_pr_payload.py
#!/usr/bin/env python3
"""
Builds the JSON payload for the typst/packages pull request.

Reads the relevant CHANGELOG.md section (if present) and writes a JSON file
to /tmp/pr_payload.json suitable for `gh api repos/typst/packages/pulls
--input /tmp/pr_payload.json`.

Environment variables (required):
  PACKAGE_NAME  e.g. cgc-resume-starter
  VERSION       e.g. 2.0.0
  BRANCH        e.g. add/cgc-resume-starter-2.0.0
"""

import json
import os
import re
import sys


def extract_changelog_entry(package_name: str, version: str) -> str:
    changelog_path = f"packages/{package_name}/CHANGELOG.md"
    if not os.path.exists(changelog_path):
        return ""
    with open(changelog_path) as f:
        content = f.read()
    # Match the heading for this version, then capture everything up to the
    # next same-level heading (or end of file)
    m = re.search(rf"^## .*{re.escape(version)}.*$", content, re.MULTILINE)
    if not m:
        return ""
    start = m.end()
    next_m = re.search(r"^## ", content[start:], re.MULTILINE)
    end = start + next_m.start() if next_m else len(content)
    return content[start:end].strip()


def main() -> None:
    package_name = os.environ.get("PACKAGE_NAME", "")
    version = os.environ.get("VERSION", "")
    branch = os.environ.get("BRANCH", "")

    if not all([package_name, version, branch]):
        print("::error::PACKAGE_NAME, VERSION, and BRANCH must be set", file=sys.stderr)
        sys.exit(1)

    body = extract_changelog_entry(package_name, version)

    payload = {
        "title": f"packages/preview/{package_name}/{version}",
        "head": f"chaoticgoodcomputing:{branch}",
        "base": "main",
        "body": body,
    }

    with open("/tmp/pr_payload.json", "w") as f:
        json.dump(payload, f)

    print(f"PR payload written to /tmp/pr_payload.json")
    print(f"  title : {payload['title']}")
    print(f"  head  : {payload['head']}")
    print(f"  body  : {len(body)} chars")


if __name__ == "__main__":
    main()
